sdk/clientcore/http-netty4/src/main/java/io/clientcore/http/netty4/implementation/Netty4ChannelInputStream.java

```diff
@@ -0,0 +1,225 @@
+
+        // currentBuffer has enough data remaining to satisfy the read.
+        if (remainingInCurrentBuffer >= n) {
+            readIndex += (int) n;
+            return n;
+        }
+
+        // Set up a marker to determine how much more to skip.
+        long toSkip = n - remainingInCurrentBuffer;
+
+        while (setupNextBuffer()) {
+            if (currentBuffer.length >= toSkip) {
+                // The next buffer is able to satisfy the remaining skip.
+                readIndex = (int) toSkip;
+                return n;
+            } else {
+                // The entire next buffer is able to fit in the remaining skip amount.
+                toSkip -= currentBuffer.length;
+            }
+        }
+
+        // Reached the end of the Channel contents before the skip amount could be satisfied.
+        // Return the amount of content skipped, which would be the skip amount requested minus what couldn't be
+        // skipped.
+        return n - toSkip;
+    }
+
+    @Override
+    public void close() {
+        currentBuffer = null;
+        additionalBuffers.clear();
+        if (!channelDone) {
+            channel.close();
+        }
+    }
+
+    private boolean setupNextBuffer() {
+        if (!additionalBuffers.isEmpty()) {
+            currentBuffer = additionalBuffers.pop();
+            readIndex = 0;
+            return true;
+        } else if (readMore()) {
+            return true;
+        } else {
+            streamDone = true;
+            return false;
+        }
+    }
+
+    // Attempts to read more data from the Channel.
+    // If more data was read true will be returned, otherwise false will be returned indicating the Channel has no more
+    // data is fully read.
+    // This method should only be called when 'additionalBuffers' is empty and 'currentBuffer' has been fully consumed.
+    private boolean readMore() {
+        // Channel has been fully read, cannot retrieve anymore data from it.
+        if (channelDone) {
+            return false;
+        }
+
+        // Check if the Channel is still active before attempting to retrieve more data.
+        if (!channel.isActive()) {
+            channelDone = true;
+            return false;
+        }
+
+        // Run reading the Channel in a loop, just in case all reads return empty data but the Channel doesn't complete.
+        while (additionalBuffers.isEmpty() && !channelDone) {
+            CountDownLatch latch = new CountDownLatch(1);
+            Netty4InitiateOneReadHandler handler = new Netty4InitiateOneReadHandler(latch, byteBuf -> {
+                // No need to check if the ByteBuf is readable as that is handled by Netty4InitiateOneReadHandler's
+                // channelRead method.
+                byte[] buffer = new byte[byteBuf.readableBytes()];
+                byteBuf.readBytes(buffer);
+
+                additionalBuffers.add(buffer);
+            });
+            channel.pipeline().addLast(handler);
```

Comment:
Is it by-design that we keep the previous handlers to channel when new handlers are added in future calls of `readMore`?

Reply:
thanks! I was also curious, do we have a mechanism in place to timeout in case `Netty4InitiateOneReadHandler ` never responds so that the `InputStream.read` won't block forever on latch. I see we added read-write-timeout handler at init time, but wanted to check if that somehow triggers the countdown. (Otherwise, maybe latch await overload with timeout could be an option to exit read await with an error.. )

Reply:
Actually, I'll make this a follow-up as there are some questions around handling when the handler is added to a closed channel that'll need further discussion.

https://github.com/Azure/azure-sdk-for-java/issues/45046